From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Stained Glass OS <ke7oxh@gmail.com>
Date: Tue, 13 Oct 2026 15:57:02 -0700
Subject: [PATCH] msxml3: XML viewer wrappers call the objects they wrap, typed
 values of plain nodes

The msxml3 XML viewer's moniker, binding, bind status callback, IOleObject, IPersistHistory, IOleCommandTarget and IPersistMoniker were E_NOTIMPL stubs that hid the aggregated HTML document. They now pass calls on to the wrapped objects. nodeTypedValue puts of unknown nodes, entity references, fragments and documents use the string value.

diff --git a/dlls/msxml3/docfrag.c b/dlls/msxml3/docfrag.c
index f8c3a0b..68cd9e1 100644
--- a/dlls/msxml3/docfrag.c
+++ b/dlls/msxml3/docfrag.c
@@ -416,8 +416,13 @@ static HRESULT WINAPI domfrag_put_nodeTypedValue(
     VARIANT typedValue)
 {
     domfrag *This = impl_from_IXMLDOMDocumentFragment( iface );
-    FIXME("(%p)->(%s)\n", This, debugstr_variant(&typedValue));
+    TRACE("(%p)->(%s)\n", This, debugstr_variant(&typedValue));
+
+    /* these nodes have no data type: the typed value is the string value */
+#ifdef SG_MUTANT_MSXML_TYPEDVAL
     return E_NOTIMPL;
+#endif
+    return IXMLDOMDocumentFragment_put_nodeValue(iface, typedValue);
 }
 
 static HRESULT WINAPI domfrag_get_dataType(
diff --git a/dlls/msxml3/domdoc.c b/dlls/msxml3/domdoc.c
index 1c3a0af..8f24ee5 100644
--- a/dlls/msxml3/domdoc.c
+++ b/dlls/msxml3/domdoc.c
@@ -1616,8 +1616,10 @@ static HRESULT WINAPI domdoc_put_nodeTypedValue(
     VARIANT typedValue )
 {
     domdoc *This = impl_from_IXMLDOMDocument3( iface );
-    FIXME("(%p)->(%s)\n", This, debugstr_variant(&typedValue));
-    return E_NOTIMPL;
+    TRACE("(%p)->(%s)\n", This, debugstr_variant(&typedValue));
+
+    /* these nodes have no data type: the typed value is the string value */
+    return IXMLDOMDocument3_put_nodeValue(iface, typedValue);
 }
 
 
diff --git a/dlls/msxml3/entityref.c b/dlls/msxml3/entityref.c
index 45a618c..d4e92a5 100644
--- a/dlls/msxml3/entityref.c
+++ b/dlls/msxml3/entityref.c
@@ -413,8 +413,10 @@ static HRESULT WINAPI entityref_put_nodeTypedValue(
     VARIANT typedValue)
 {
     entityref *This = impl_from_IXMLDOMEntityReference( iface );
-    FIXME("(%p)->(%s)\n", This, debugstr_variant(&typedValue));
-    return E_NOTIMPL;
+    TRACE("(%p)->(%s)\n", This, debugstr_variant(&typedValue));
+
+    /* these nodes have no data type: the typed value is the string value */
+    return IXMLDOMEntityReference_put_nodeValue(iface, typedValue);
 }
 
 static HRESULT WINAPI entityref_get_dataType(
diff --git a/dlls/msxml3/node.c b/dlls/msxml3/node.c
index 27348b3..4f07023 100644
--- a/dlls/msxml3/node.c
+++ b/dlls/msxml3/node.c
@@ -2346,8 +2346,10 @@ static HRESULT WINAPI unknode_put_nodeTypedValue(
     VARIANT typedValue)
 {
     unknode *This = unknode_from_IXMLDOMNode( iface );
-    FIXME("(%p)->(%s)\n", This, debugstr_variant(&typedValue));
-    return E_NOTIMPL;
+    TRACE("(%p)->(%s)\n", This, debugstr_variant(&typedValue));
+
+    /* these nodes have no data type: the typed value is the string value */
+    return IXMLDOMNode_put_nodeValue(iface, typedValue);
 }
 
 static HRESULT WINAPI unknode_get_dataType(
diff --git a/dlls/msxml3/xmlview.c b/dlls/msxml3/xmlview.c
index 0705741..01b6d81 100644
--- a/dlls/msxml3/xmlview.c
+++ b/dlls/msxml3/xmlview.c
@@ -66,6 +66,8 @@ typedef struct
     IStream *stream;
 } BindStatusCallback;
 
+static const CLSID xmlview_clsid = {0x48123bc4,0x99d9,0x11d1,{0xa6,0xb3,0x00,0xc0,0x4f,0xd9,0x15,0x55}};
+
 typedef struct
 {
     IBinding IBinding_iface;
@@ -130,15 +132,15 @@ static HRESULT WINAPI XMLView_Binding_Abort(IBinding *iface)
 static HRESULT WINAPI XMLView_Binding_Suspend(IBinding *iface)
 {
     Binding *This = impl_from_IBinding(iface);
-    FIXME("(%p)\n", This);
-    return E_NOTIMPL;
+    TRACE("(%p)\n", This);
+    return IBinding_Suspend(This->binding);
 }
 
 static HRESULT WINAPI XMLView_Binding_Resume(IBinding *iface)
 {
     Binding *This = impl_from_IBinding(iface);
-    FIXME("(%p)\n", This);
-    return E_NOTIMPL;
+    TRACE("(%p)\n", This);
+    return IBinding_Resume(This->binding);
 }
 
 static HRESULT WINAPI XMLView_Binding_SetPriority(
@@ -164,9 +166,8 @@ static HRESULT WINAPI XMLView_Binding_GetBindResult(IBinding *iface,
         DWORD *pdwReserved)
 {
     Binding *This = impl_from_IBinding(iface);
-    FIXME("(%p)->(%s %p %p %p)\n", This, debugstr_guid(pclsidProtocol),
-            pdwResult, pszResult, pdwReserved);
-    return E_NOTIMPL;
+    TRACE("(%p)->(%s %p %p %p)\n", This, debugstr_guid(pclsidProtocol), pdwResult, pszResult, pdwReserved);
+    return IBinding_GetBindResult(This->binding, pclsidProtocol, pdwResult, pszResult, pdwReserved);
 }
 
 static IBindingVtbl XMLView_BindingVtbl = {
@@ -281,15 +282,16 @@ static HRESULT WINAPI XMLView_BindStatusCallback_GetPriority(
         IBindStatusCallback *iface, LONG *pnPriority)
 {
     BindStatusCallback *This = impl_from_IBindStatusCallback(iface);
-    FIXME("(%p)->(%p)\n", This, pnPriority);
-    return E_NOTIMPL;
+    TRACE("(%p)->(%p)\n", This, pnPriority);
+    return IBindStatusCallback_GetPriority(This->bsc, pnPriority);
 }
 
 static HRESULT WINAPI XMLView_BindStatusCallback_OnLowResource(
         IBindStatusCallback *iface, DWORD reserved)
 {
-    FIXME("%p, %lx.\n", iface, reserved);
-    return E_NOTIMPL;
+    BindStatusCallback *This = impl_from_IBindStatusCallback(iface);
+    TRACE("%p, %lx.\n", iface, reserved);
+    return IBindStatusCallback_OnLowResource(This->bsc, reserved);
 }
 
 static HRESULT WINAPI XMLView_BindStatusCallback_OnProgress(
@@ -307,8 +309,8 @@ static HRESULT WINAPI XMLView_BindStatusCallback_OnProgress(
     case BINDSTATUS_MIMETYPEAVAILABLE:
         return S_OK;
     default:
-        FIXME("ulStatusCode: %lu\n", ulStatusCode);
-        return E_NOTIMPL;
+        /* the other progress reports are not for the viewer: the host's callback shows them */
+        return IBindStatusCallback_OnProgress(This->bsc, ulProgress, ulProgressMax, ulStatusCode, szStatusText);
     }
 }
 
@@ -532,8 +534,8 @@ static HRESULT WINAPI XMLView_BindStatusCallback_OnObjectAvailable(
         IBindStatusCallback *iface, REFIID riid, IUnknown *punk)
 {
     BindStatusCallback *This = impl_from_IBindStatusCallback(iface);
-    FIXME("(%p)->(%s %p)\n", This, debugstr_guid(riid), punk);
-    return E_NOTIMPL;
+    TRACE("(%p)->(%s %p)\n", This, debugstr_guid(riid), punk);
+    return IBindStatusCallback_OnObjectAvailable(This->bsc, riid, punk);
 }
 
 static IBindStatusCallbackVtbl XMLView_BindStatusCallbackVtbl = {
@@ -623,47 +625,46 @@ static ULONG WINAPI XMLView_Moniker_Release(IMoniker *iface)
 static HRESULT WINAPI XMLView_Moniker_GetClassID(IMoniker *iface, CLSID *pClassID)
 {
     Moniker *This = impl_from_IMoniker(iface);
-    FIXME("(%p)->(%p)\n", This, pClassID);
-    return E_NOTIMPL;
+    TRACE("(%p)->(%p)\n", This, pClassID);
+    return IMoniker_GetClassID(This->mon, pClassID);
 }
 
 static HRESULT WINAPI XMLView_Moniker_IsDirty(IMoniker *iface)
 {
     Moniker *This = impl_from_IMoniker(iface);
-    FIXME("(%p)\n", This);
-    return E_NOTIMPL;
+    TRACE("(%p)\n", This);
+    return IMoniker_IsDirty(This->mon);
 }
 
 static HRESULT WINAPI XMLView_Moniker_Load(IMoniker *iface, IStream *pStm)
 {
     Moniker *This = impl_from_IMoniker(iface);
-    FIXME("(%p)->(%p)\n", This, pStm);
-    return E_NOTIMPL;
+    TRACE("(%p)->(%p)\n", This, pStm);
+    return IMoniker_Load(This->mon, pStm);
 }
 
 static HRESULT WINAPI XMLView_Moniker_Save(IMoniker *iface,
         IStream *pStm, BOOL fClearDirty)
 {
     Moniker *This = impl_from_IMoniker(iface);
-    FIXME("(%p)->(%p %x)\n", This, pStm, fClearDirty);
-    return E_NOTIMPL;
+    TRACE("(%p)->(%p %x)\n", This, pStm, fClearDirty);
+    return IMoniker_Save(This->mon, pStm, fClearDirty);
 }
 
 static HRESULT WINAPI XMLView_Moniker_GetSizeMax(
         IMoniker *iface, ULARGE_INTEGER *pcbSize)
 {
     Moniker *This = impl_from_IMoniker(iface);
-    FIXME("(%p)->(%p)\n", This, pcbSize);
-    return E_NOTIMPL;
+    TRACE("(%p)->(%p)\n", This, pcbSize);
+    return IMoniker_GetSizeMax(This->mon, pcbSize);
 }
 
 static HRESULT WINAPI XMLView_Moniker_BindToObject(IMoniker *iface, IBindCtx *pbc,
         IMoniker *pmkToLeft, REFIID riidResult, void **ppvResult)
 {
     Moniker *This = impl_from_IMoniker(iface);
-    FIXME("(%p)->(%p %p %s %p)\n", This, pbc, pmkToLeft,
-            debugstr_guid(riidResult), ppvResult);
-    return E_NOTIMPL;
+    TRACE("(%p)->(%p %p %s %p)\n", This, pbc, pmkToLeft, debugstr_guid(riidResult), ppvResult);
+    return IMoniker_BindToObject(This->mon, pbc, pmkToLeft, riidResult, ppvResult);
 }
 
 static HRESULT WINAPI XMLView_Moniker_BindToStorage(IMoniker *iface, IBindCtx *pbc,
@@ -688,78 +689,79 @@ static HRESULT WINAPI XMLView_Moniker_BindToStorage(IMoniker *iface, IBindCtx *p
 static HRESULT WINAPI XMLView_Moniker_Reduce(IMoniker *iface, IBindCtx *pbc,
         DWORD dwReduceHowFar, IMoniker **ppmkToLeft, IMoniker **ppmkReduced)
 {
-    FIXME("%p, %p, %ld, %p, %p.\n", iface, pbc, dwReduceHowFar, ppmkToLeft, ppmkReduced);
-    return E_NOTIMPL;
+    Moniker *This = impl_from_IMoniker(iface);
+    TRACE("%p, %p, %ld, %p, %p.\n", iface, pbc, dwReduceHowFar, ppmkToLeft, ppmkReduced);
+    return IMoniker_Reduce(This->mon, pbc, dwReduceHowFar, ppmkToLeft, ppmkReduced);
 }
 
 static HRESULT WINAPI XMLView_Moniker_ComposeWith(IMoniker *iface, IMoniker *pmkRight,
         BOOL fOnlyIfNotGeneric, IMoniker **ppmkComposite)
 {
     Moniker *This = impl_from_IMoniker(iface);
-    FIXME("(%p)->(%p %x %p)\n", This, pmkRight, fOnlyIfNotGeneric, ppmkComposite);
-    return E_NOTIMPL;
+    TRACE("(%p)->(%p %x %p)\n", This, pmkRight, fOnlyIfNotGeneric, ppmkComposite);
+    return IMoniker_ComposeWith(This->mon, pmkRight, fOnlyIfNotGeneric, ppmkComposite);
 }
 
 static HRESULT WINAPI XMLView_Moniker_Enum(IMoniker *iface,
         BOOL fForward, IEnumMoniker **ppenumMoniker)
 {
     Moniker *This = impl_from_IMoniker(iface);
-    FIXME("(%p)->(%x %p)\n", This, fForward, ppenumMoniker);
-    return E_NOTIMPL;
+    TRACE("(%p)->(%x %p)\n", This, fForward, ppenumMoniker);
+    return IMoniker_Enum(This->mon, fForward, ppenumMoniker);
 }
 
 static HRESULT WINAPI XMLView_Moniker_IsEqual(IMoniker *iface,
         IMoniker *pmkOtherMoniker)
 {
     Moniker *This = impl_from_IMoniker(iface);
-    FIXME("(%p)->(%p)\n", This, pmkOtherMoniker);
-    return E_NOTIMPL;
+    TRACE("(%p)->(%p)\n", This, pmkOtherMoniker);
+    return IMoniker_IsEqual(This->mon, pmkOtherMoniker);
 }
 
 static HRESULT WINAPI XMLView_Moniker_Hash(IMoniker *iface, DWORD *pdwHash)
 {
     Moniker *This = impl_from_IMoniker(iface);
-    FIXME("(%p)->(%p)\n", This, pdwHash);
-    return E_NOTIMPL;
+    TRACE("(%p)->(%p)\n", This, pdwHash);
+    return IMoniker_Hash(This->mon, pdwHash);
 }
 
 static HRESULT WINAPI XMLView_Moniker_IsRunning(IMoniker *iface, IBindCtx *pbc,
         IMoniker *pmkToLeft, IMoniker *pmkNewlyRunning)
 {
     Moniker *This = impl_from_IMoniker(iface);
-    FIXME("(%p)->(%p %p %p)\n", This, pbc, pmkToLeft, pmkNewlyRunning);
-    return E_NOTIMPL;
+    TRACE("(%p)->(%p %p %p)\n", This, pbc, pmkToLeft, pmkNewlyRunning);
+    return IMoniker_IsRunning(This->mon, pbc, pmkToLeft, pmkNewlyRunning);
 }
 
 static HRESULT WINAPI XMLView_Moniker_GetTimeOfLastChange(IMoniker *iface,
         IBindCtx *pbc, IMoniker *pmkToLeft, FILETIME *pFileTime)
 {
     Moniker *This = impl_from_IMoniker(iface);
-    FIXME("(%p)->(%p %p %p)\n", This, pbc, pmkToLeft, pFileTime);
-    return E_NOTIMPL;
+    TRACE("(%p)->(%p %p %p)\n", This, pbc, pmkToLeft, pFileTime);
+    return IMoniker_GetTimeOfLastChange(This->mon, pbc, pmkToLeft, pFileTime);
 }
 
 static HRESULT WINAPI XMLView_Moniker_Inverse(IMoniker *iface, IMoniker **ppmk)
 {
     Moniker *This = impl_from_IMoniker(iface);
-    FIXME("(%p)->(%p)\n", This, ppmk);
-    return E_NOTIMPL;
+    TRACE("(%p)->(%p)\n", This, ppmk);
+    return IMoniker_Inverse(This->mon, ppmk);
 }
 
 static HRESULT WINAPI XMLView_Moniker_CommonPrefixWith(IMoniker *iface,
         IMoniker *pmkOther, IMoniker **ppmkPrefix)
 {
     Moniker *This = impl_from_IMoniker(iface);
-    FIXME("(%p)->(%p %p)\n", This, pmkOther, ppmkPrefix);
-    return E_NOTIMPL;
+    TRACE("(%p)->(%p %p)\n", This, pmkOther, ppmkPrefix);
+    return IMoniker_CommonPrefixWith(This->mon, pmkOther, ppmkPrefix);
 }
 
 static HRESULT WINAPI XMLView_Moniker_RelativePathTo(IMoniker *iface,
         IMoniker *pmkOther, IMoniker **ppmkRelPath)
 {
     Moniker *This = impl_from_IMoniker(iface);
-    FIXME("(%p)->(%p %p)\n", This, pmkOther, ppmkRelPath);
-    return E_NOTIMPL;
+    TRACE("(%p)->(%p %p)\n", This, pmkOther, ppmkRelPath);
+    return IMoniker_RelativePathTo(This->mon, pmkOther, ppmkRelPath);
 }
 
 static HRESULT WINAPI XMLView_Moniker_GetDisplayName(IMoniker *iface,
@@ -775,17 +777,16 @@ static HRESULT WINAPI XMLView_Moniker_ParseDisplayName(IMoniker *iface,
         ULONG *pchEaten, IMoniker **ppmkOut)
 {
     Moniker *This = impl_from_IMoniker(iface);
-    FIXME("(%p)->(%p %p %s %p %p)\n", This, pbc, pmkToLeft,
-            debugstr_w(pszDisplayName), pchEaten, ppmkOut);
-    return E_NOTIMPL;
+    TRACE("(%p)->(%p %p %s %p %p)\n", This, pbc, pmkToLeft, debugstr_w(pszDisplayName), pchEaten, ppmkOut);
+    return IMoniker_ParseDisplayName(This->mon, pbc, pmkToLeft, pszDisplayName, pchEaten, ppmkOut);
 }
 
 static HRESULT WINAPI XMLView_Moniker_IsSystemMoniker(
         IMoniker *iface, DWORD *pdwMksys)
 {
     Moniker *This = impl_from_IMoniker(iface);
-    FIXME("(%p)->(%p)\n", This, pdwMksys);
-    return E_NOTIMPL;
+    TRACE("(%p)->(%p)\n", This, pdwMksys);
+    return IMoniker_IsSystemMoniker(This->mon, pdwMksys);
 }
 
 static IMonikerVtbl XMLView_MonikerVtbl = {
@@ -893,15 +894,29 @@ static HRESULT WINAPI XMLView_PersistMoniker_GetClassID(
         IPersistMoniker *iface, CLSID *pClassID)
 {
     XMLView *This = impl_from_IPersistMoniker(iface);
-    FIXME("(%p)->(%p)\n", This, pClassID);
-    return E_NOTIMPL;
+    TRACE("(%p)->(%p)\n", This, pClassID);
+
+    if(!pClassID)
+        return E_POINTER;
+    *pClassID = xmlview_clsid;
+    return S_OK;
 }
 
 static HRESULT WINAPI XMLView_PersistMoniker_IsDirty(IPersistMoniker *iface)
 {
     XMLView *This = impl_from_IPersistMoniker(iface);
-    FIXME("(%p)\n", This);
-    return E_NOTIMPL;
+    IPersistMoniker *inner;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    /* the viewer is the HTML document in front of the transformed XML: the call is for that document */
+    hres = IUnknown_QueryInterface(This->html_doc, &IID_IPersistMoniker, (void**)&inner);
+    if(FAILED(hres))
+        return hres;
+    hres = IPersistMoniker_IsDirty(inner);
+    IPersistMoniker_Release(inner);
+    return hres;
 }
 
 static HRESULT WINAPI XMLView_PersistMoniker_Load(IPersistMoniker *iface,
@@ -1038,24 +1053,51 @@ static HRESULT WINAPI XMLView_PersistMoniker_Save(IPersistMoniker *iface,
         IMoniker *pimkName, LPBC pbc, BOOL fRemember)
 {
     XMLView *This = impl_from_IPersistMoniker(iface);
-    FIXME("(%p)->(%p %p %x)\n", This, pimkName, pbc, fRemember);
-    return E_NOTIMPL;
+    IPersistMoniker *inner;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    /* the viewer is the HTML document in front of the transformed XML: the call is for that document */
+    hres = IUnknown_QueryInterface(This->html_doc, &IID_IPersistMoniker, (void**)&inner);
+    if(FAILED(hres))
+        return hres;
+    hres = IPersistMoniker_Save(inner, pimkName, pbc, fRemember);
+    IPersistMoniker_Release(inner);
+    return hres;
 }
 
 static HRESULT WINAPI XMLView_PersistMoniker_SaveCompleted(
         IPersistMoniker *iface, IMoniker *pimkName, LPBC pibc)
 {
     XMLView *This = impl_from_IPersistMoniker(iface);
-    FIXME("(%p)->(%p %p)\n", This, pimkName, pibc);
-    return E_NOTIMPL;
+    IPersistMoniker *inner;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    /* the viewer is the HTML document in front of the transformed XML: the call is for that document */
+    hres = IUnknown_QueryInterface(This->html_doc, &IID_IPersistMoniker, (void**)&inner);
+    if(FAILED(hres))
+        return hres;
+    hres = IPersistMoniker_SaveCompleted(inner, pimkName, pibc);
+    IPersistMoniker_Release(inner);
+    return hres;
 }
 
 static HRESULT WINAPI XMLView_PersistMoniker_GetCurMoniker(
         IPersistMoniker *iface, IMoniker **ppimkName)
 {
     XMLView *This = impl_from_IPersistMoniker(iface);
-    FIXME("(%p)->(%p)\n", This, ppimkName);
-    return E_NOTIMPL;
+    TRACE("(%p)->(%p)\n", This, ppimkName);
+
+    if(!ppimkName)
+        return E_POINTER;
+    *ppimkName = This->mon;
+    if(!This->mon)
+        return E_UNEXPECTED;
+    IMoniker_AddRef(This->mon);
+    return S_OK;
 }
 
 static IPersistMonikerVtbl XMLView_PersistMonikerVtbl = {
@@ -1098,39 +1140,90 @@ static HRESULT WINAPI XMLView_PersistHistory_GetClassID(
         IPersistHistory *iface, CLSID *pClassID)
 {
     XMLView *This = impl_from_IPersistHistory(iface);
-    FIXME("(%p)->(%p)\n", This, pClassID);
-    return E_NOTIMPL;
+    IPersistHistory *inner;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    /* the viewer is the HTML document in front of the transformed XML: the call is for that document */
+    hres = IUnknown_QueryInterface(This->html_doc, &IID_IPersistHistory, (void**)&inner);
+    if(FAILED(hres))
+        return hres;
+    hres = IPersistHistory_GetClassID(inner, pClassID);
+    IPersistHistory_Release(inner);
+    return hres;
 }
 
 static HRESULT WINAPI XMLView_PersistHistory_LoadHistory(
         IPersistHistory *iface, IStream *pStream, IBindCtx *pbc)
 {
     XMLView *This = impl_from_IPersistHistory(iface);
-    FIXME("(%p)->(%p %p)\n", This, pStream, pbc);
-    return E_NOTIMPL;
+    IPersistHistory *inner;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    /* the viewer is the HTML document in front of the transformed XML: the call is for that document */
+    hres = IUnknown_QueryInterface(This->html_doc, &IID_IPersistHistory, (void**)&inner);
+    if(FAILED(hres))
+        return hres;
+    hres = IPersistHistory_LoadHistory(inner, pStream, pbc);
+    IPersistHistory_Release(inner);
+    return hres;
 }
 
 static HRESULT WINAPI XMLView_PersistHistory_SaveHistory(
         IPersistHistory *iface, IStream *pStream)
 {
     XMLView *This = impl_from_IPersistHistory(iface);
-    FIXME("(%p)->(%p)\n", This, pStream);
-    return E_NOTIMPL;
+    IPersistHistory *inner;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    /* the viewer is the HTML document in front of the transformed XML: the call is for that document */
+    hres = IUnknown_QueryInterface(This->html_doc, &IID_IPersistHistory, (void**)&inner);
+    if(FAILED(hres))
+        return hres;
+    hres = IPersistHistory_SaveHistory(inner, pStream);
+    IPersistHistory_Release(inner);
+    return hres;
 }
 
 static HRESULT WINAPI XMLView_PersistHistory_SetPositionCookie(
         IPersistHistory *iface, DWORD dwPositioncookie)
 {
-    FIXME("%p, %ld.\n", iface, dwPositioncookie);
-    return E_NOTIMPL;
+    XMLView *This = impl_from_IPersistHistory(iface);
+    IPersistHistory *inner;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    /* the viewer is the HTML document in front of the transformed XML: the call is for that document */
+    hres = IUnknown_QueryInterface(This->html_doc, &IID_IPersistHistory, (void**)&inner);
+    if(FAILED(hres))
+        return hres;
+    hres = IPersistHistory_SetPositionCookie(inner, dwPositioncookie);
+    IPersistHistory_Release(inner);
+    return hres;
 }
 
 static HRESULT WINAPI XMLView_PersistHistory_GetPositionCookie(
         IPersistHistory *iface, DWORD *pdwPositioncookie)
 {
     XMLView *This = impl_from_IPersistHistory(iface);
-    FIXME("(%p)->(%p)\n", This, pdwPositioncookie);
-    return E_NOTIMPL;
+    IPersistHistory *inner;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    /* the viewer is the HTML document in front of the transformed XML: the call is for that document */
+    hres = IUnknown_QueryInterface(This->html_doc, &IID_IPersistHistory, (void**)&inner);
+    if(FAILED(hres))
+        return hres;
+    hres = IPersistHistory_GetPositionCookie(inner, pdwPositioncookie);
+    IPersistHistory_Release(inner);
+    return hres;
 }
 
 static IPersistHistoryVtbl XMLView_PersistHistoryVtbl = {
@@ -1171,17 +1264,38 @@ static ULONG WINAPI XMLView_OleCommandTarget_Release(IOleCommandTarget *iface)
 static HRESULT WINAPI XMLView_OleCommandTarget_QueryStatus(IOleCommandTarget *iface,
         const GUID *pguidCmdGroup, ULONG cCmds, OLECMD prgCmds[], OLECMDTEXT *pCmdText)
 {
-    FIXME("%p, %p, %lu, %p, %p.\n", iface, pguidCmdGroup, cCmds, prgCmds, pCmdText);
-    return E_NOTIMPL;
+    XMLView *This = impl_from_IOleCommandTarget(iface);
+    IOleCommandTarget *inner;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    /* the viewer is the HTML document in front of the transformed XML: the call is for that document */
+    hres = IUnknown_QueryInterface(This->html_doc, &IID_IOleCommandTarget, (void**)&inner);
+    if(FAILED(hres))
+        return hres;
+    hres = IOleCommandTarget_QueryStatus(inner, pguidCmdGroup, cCmds, prgCmds, pCmdText);
+    IOleCommandTarget_Release(inner);
+    return hres;
 }
 
 static HRESULT WINAPI XMLView_OleCommandTarget_Exec(IOleCommandTarget *iface,
         const GUID *pguidCmdGroup, DWORD nCmdID, DWORD nCmdexecopt,
         VARIANT *pvaIn, VARIANT *pvaOut)
 {
-    FIXME("%p, %p, %ld, %lx, %p, %p.\n", iface, pguidCmdGroup,
-            nCmdID, nCmdexecopt, pvaIn, pvaOut);
-    return E_NOTIMPL;
+    XMLView *This = impl_from_IOleCommandTarget(iface);
+    IOleCommandTarget *inner;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    /* the viewer is the HTML document in front of the transformed XML: the call is for that document */
+    hres = IUnknown_QueryInterface(This->html_doc, &IID_IOleCommandTarget, (void**)&inner);
+    if(FAILED(hres))
+        return hres;
+    hres = IOleCommandTarget_Exec(inner, pguidCmdGroup, nCmdID, nCmdexecopt, pvaIn, pvaOut);
+    IOleCommandTarget_Release(inner);
+    return hres;
 }
 
 static IOleCommandTargetVtbl XMLView_OleCommandTargetVtbl = {
@@ -1220,155 +1334,379 @@ static HRESULT WINAPI XMLView_OleObject_SetClientSite(
         IOleObject *iface, IOleClientSite *pClientSite)
 {
     XMLView *This = impl_from_IOleObject(iface);
-    FIXME("(%p)->(%p)\n", This, pClientSite);
-    return E_NOTIMPL;
+    IOleObject *inner;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    /* the viewer is the HTML document in front of the transformed XML: the call is for that document */
+    hres = IUnknown_QueryInterface(This->html_doc, &IID_IOleObject, (void**)&inner);
+    if(FAILED(hres))
+        return hres;
+    hres = IOleObject_SetClientSite(inner, pClientSite);
+    IOleObject_Release(inner);
+    return hres;
 }
 
 static HRESULT WINAPI XMLView_OleObject_GetClientSite(
         IOleObject *iface, IOleClientSite **ppClientSite)
 {
     XMLView *This = impl_from_IOleObject(iface);
-    FIXME("(%p)->(%p)\n", This, ppClientSite);
-    return E_NOTIMPL;
+    IOleObject *inner;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    /* the viewer is the HTML document in front of the transformed XML: the call is for that document */
+    hres = IUnknown_QueryInterface(This->html_doc, &IID_IOleObject, (void**)&inner);
+    if(FAILED(hres))
+        return hres;
+    hres = IOleObject_GetClientSite(inner, ppClientSite);
+    IOleObject_Release(inner);
+    return hres;
 }
 
 static HRESULT WINAPI XMLView_OleObject_SetHostNames(IOleObject *iface,
         LPCOLESTR szContainerApp, LPCOLESTR szContainerObj)
 {
     XMLView *This = impl_from_IOleObject(iface);
-    FIXME("(%p)->(%s %s)\n", This, debugstr_w(szContainerApp), debugstr_w(szContainerObj));
-    return E_NOTIMPL;
+    IOleObject *inner;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    /* the viewer is the HTML document in front of the transformed XML: the call is for that document */
+    hres = IUnknown_QueryInterface(This->html_doc, &IID_IOleObject, (void**)&inner);
+    if(FAILED(hres))
+        return hres;
+    hres = IOleObject_SetHostNames(inner, szContainerApp, szContainerObj);
+    IOleObject_Release(inner);
+    return hres;
 }
 
 static HRESULT WINAPI XMLView_OleObject_Close(IOleObject *iface, DWORD dwSaveOption)
 {
-    FIXME("%p, %lx.\n", iface, dwSaveOption);
-    return E_NOTIMPL;
+    XMLView *This = impl_from_IOleObject(iface);
+    IOleObject *inner;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    /* the viewer is the HTML document in front of the transformed XML: the call is for that document */
+    hres = IUnknown_QueryInterface(This->html_doc, &IID_IOleObject, (void**)&inner);
+    if(FAILED(hres))
+        return hres;
+    hres = IOleObject_Close(inner, dwSaveOption);
+    IOleObject_Release(inner);
+    return hres;
 }
 
 static HRESULT WINAPI XMLView_OleObject_SetMoniker(IOleObject *iface,
         DWORD dwWhichMoniker, IMoniker *pmk)
 {
-    FIXME("%p, %lx, %p.\n", iface, dwWhichMoniker, pmk);
-    return E_NOTIMPL;
+    XMLView *This = impl_from_IOleObject(iface);
+    IOleObject *inner;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    /* the viewer is the HTML document in front of the transformed XML: the call is for that document */
+    hres = IUnknown_QueryInterface(This->html_doc, &IID_IOleObject, (void**)&inner);
+    if(FAILED(hres))
+        return hres;
+    hres = IOleObject_SetMoniker(inner, dwWhichMoniker, pmk);
+    IOleObject_Release(inner);
+    return hres;
 }
 
 static HRESULT WINAPI XMLView_OleObject_GetMoniker(IOleObject *iface,
         DWORD dwAssign, DWORD dwWhichMoniker, IMoniker **ppmk)
 {
-    FIXME("%p, %lx, %lx, %p.\n", iface, dwAssign, dwWhichMoniker, ppmk);
-    return E_NOTIMPL;
+    XMLView *This = impl_from_IOleObject(iface);
+    IOleObject *inner;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    /* the viewer is the HTML document in front of the transformed XML: the call is for that document */
+    hres = IUnknown_QueryInterface(This->html_doc, &IID_IOleObject, (void**)&inner);
+    if(FAILED(hres))
+        return hres;
+    hres = IOleObject_GetMoniker(inner, dwAssign, dwWhichMoniker, ppmk);
+    IOleObject_Release(inner);
+    return hres;
 }
 
 static HRESULT WINAPI XMLView_OleObject_InitFromData(IOleObject *iface,
         IDataObject *pDataObject, BOOL fCreation, DWORD dwReserved)
 {
-    FIXME("%p, %p, %x, %lx.\n", iface, pDataObject, fCreation, dwReserved);
-    return E_NOTIMPL;
+    XMLView *This = impl_from_IOleObject(iface);
+    IOleObject *inner;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    /* the viewer is the HTML document in front of the transformed XML: the call is for that document */
+    hres = IUnknown_QueryInterface(This->html_doc, &IID_IOleObject, (void**)&inner);
+    if(FAILED(hres))
+        return hres;
+    hres = IOleObject_InitFromData(inner, pDataObject, fCreation, dwReserved);
+    IOleObject_Release(inner);
+    return hres;
 }
 
 static HRESULT WINAPI XMLView_OleObject_GetClipboardData(IOleObject *iface,
         DWORD dwReserved, IDataObject **ppDataObject)
 {
-    FIXME("%p, %lx, %p.\n", iface, dwReserved, ppDataObject);
-    return E_NOTIMPL;
+    XMLView *This = impl_from_IOleObject(iface);
+    IOleObject *inner;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    /* the viewer is the HTML document in front of the transformed XML: the call is for that document */
+    hres = IUnknown_QueryInterface(This->html_doc, &IID_IOleObject, (void**)&inner);
+    if(FAILED(hres))
+        return hres;
+    hres = IOleObject_GetClipboardData(inner, dwReserved, ppDataObject);
+    IOleObject_Release(inner);
+    return hres;
 }
 
 static HRESULT WINAPI XMLView_OleObject_DoVerb(IOleObject *iface,
         LONG iVerb, LPMSG lpmsg, IOleClientSite *pActiveSite,
         LONG lindex, HWND hwndParent, LPCRECT lprcPosRect)
 {
-    FIXME("%p, %ld, %p, %p, %ld, %p, %p.\n", iface, iVerb, lpmsg,
-            pActiveSite, lindex, hwndParent, lprcPosRect);
-    return E_NOTIMPL;
+    XMLView *This = impl_from_IOleObject(iface);
+    IOleObject *inner;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    /* the viewer is the HTML document in front of the transformed XML: the call is for that document */
+    hres = IUnknown_QueryInterface(This->html_doc, &IID_IOleObject, (void**)&inner);
+    if(FAILED(hres))
+        return hres;
+    hres = IOleObject_DoVerb(inner, iVerb, lpmsg, pActiveSite, lindex, hwndParent, lprcPosRect);
+    IOleObject_Release(inner);
+    return hres;
 }
 
 static HRESULT WINAPI XMLView_OleObject_EnumVerbs(
         IOleObject *iface, IEnumOLEVERB **ppEnumOleVerb)
 {
     XMLView *This = impl_from_IOleObject(iface);
-    FIXME("(%p)->{%p)\n", This, ppEnumOleVerb);
-    return E_NOTIMPL;
+    IOleObject *inner;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    /* the viewer is the HTML document in front of the transformed XML: the call is for that document */
+    hres = IUnknown_QueryInterface(This->html_doc, &IID_IOleObject, (void**)&inner);
+    if(FAILED(hres))
+        return hres;
+    hres = IOleObject_EnumVerbs(inner, ppEnumOleVerb);
+    IOleObject_Release(inner);
+    return hres;
 }
 
 static HRESULT WINAPI XMLView_OleObject_Update(IOleObject *iface)
 {
     XMLView *This = impl_from_IOleObject(iface);
-    FIXME("(%p)\n", This);
-    return E_NOTIMPL;
+    IOleObject *inner;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    /* the viewer is the HTML document in front of the transformed XML: the call is for that document */
+    hres = IUnknown_QueryInterface(This->html_doc, &IID_IOleObject, (void**)&inner);
+    if(FAILED(hres))
+        return hres;
+    hres = IOleObject_Update(inner);
+    IOleObject_Release(inner);
+    return hres;
 }
 
 static HRESULT WINAPI XMLView_OleObject_IsUpToDate(IOleObject *iface)
 {
     XMLView *This = impl_from_IOleObject(iface);
-    FIXME("(%p)\n", This);
-    return E_NOTIMPL;
+    IOleObject *inner;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    /* the viewer is the HTML document in front of the transformed XML: the call is for that document */
+    hres = IUnknown_QueryInterface(This->html_doc, &IID_IOleObject, (void**)&inner);
+    if(FAILED(hres))
+        return hres;
+    hres = IOleObject_IsUpToDate(inner);
+    IOleObject_Release(inner);
+    return hres;
 }
 
 static HRESULT WINAPI XMLView_OleObject_GetUserClassID(IOleObject *iface, CLSID *pClsid)
 {
     XMLView *This = impl_from_IOleObject(iface);
-    FIXME("(%p)->(%p)\n", This, pClsid);
-    return E_NOTIMPL;
+    IOleObject *inner;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    /* the viewer is the HTML document in front of the transformed XML: the call is for that document */
+    hres = IUnknown_QueryInterface(This->html_doc, &IID_IOleObject, (void**)&inner);
+    if(FAILED(hres))
+        return hres;
+    hres = IOleObject_GetUserClassID(inner, pClsid);
+    IOleObject_Release(inner);
+    return hres;
 }
 
 static HRESULT WINAPI XMLView_OleObject_GetUserType(IOleObject *iface,
         DWORD dwFormOfType, LPOLESTR *pszUserType)
 {
-    FIXME("%p, %lx, %p.\n", iface, dwFormOfType, pszUserType);
-    return E_NOTIMPL;
+    XMLView *This = impl_from_IOleObject(iface);
+    IOleObject *inner;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    /* the viewer is the HTML document in front of the transformed XML: the call is for that document */
+    hres = IUnknown_QueryInterface(This->html_doc, &IID_IOleObject, (void**)&inner);
+    if(FAILED(hres))
+        return hres;
+    hres = IOleObject_GetUserType(inner, dwFormOfType, pszUserType);
+    IOleObject_Release(inner);
+    return hres;
 }
 
 static HRESULT WINAPI XMLView_OleObject_SetExtent(IOleObject *iface,
         DWORD dwDrawAspect, SIZEL *psizel)
 {
-    FIXME("%p, %lx, %p.\n", iface, dwDrawAspect, psizel);
-    return E_NOTIMPL;
+    XMLView *This = impl_from_IOleObject(iface);
+    IOleObject *inner;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    /* the viewer is the HTML document in front of the transformed XML: the call is for that document */
+    hres = IUnknown_QueryInterface(This->html_doc, &IID_IOleObject, (void**)&inner);
+    if(FAILED(hres))
+        return hres;
+    hres = IOleObject_SetExtent(inner, dwDrawAspect, psizel);
+    IOleObject_Release(inner);
+    return hres;
 }
 
 static HRESULT WINAPI XMLView_OleObject_GetExtent(IOleObject *iface,
         DWORD dwDrawAspect, SIZEL *psizel)
 {
-    FIXME("%p, %lx, %p.\n", iface, dwDrawAspect, psizel);
-    return E_NOTIMPL;
+    XMLView *This = impl_from_IOleObject(iface);
+    IOleObject *inner;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    /* the viewer is the HTML document in front of the transformed XML: the call is for that document */
+    hres = IUnknown_QueryInterface(This->html_doc, &IID_IOleObject, (void**)&inner);
+    if(FAILED(hres))
+        return hres;
+    hres = IOleObject_GetExtent(inner, dwDrawAspect, psizel);
+    IOleObject_Release(inner);
+    return hres;
 }
 
 static HRESULT WINAPI XMLView_OleObject_Advise(IOleObject *iface,
         IAdviseSink *pAdvSink, DWORD *pdwConnection)
 {
     XMLView *This = impl_from_IOleObject(iface);
-    FIXME("(%p)->(%p %p)\n", This, pAdvSink, pdwConnection);
-    return E_NOTIMPL;
+    IOleObject *inner;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    /* the viewer is the HTML document in front of the transformed XML: the call is for that document */
+    hres = IUnknown_QueryInterface(This->html_doc, &IID_IOleObject, (void**)&inner);
+    if(FAILED(hres))
+        return hres;
+    hres = IOleObject_Advise(inner, pAdvSink, pdwConnection);
+    IOleObject_Release(inner);
+    return hres;
 }
 
 static HRESULT WINAPI XMLView_OleObject_Unadvise(
         IOleObject *iface, DWORD dwConnection)
 {
-    FIXME("%p, %ld.\n", iface, dwConnection);
-    return E_NOTIMPL;
+    XMLView *This = impl_from_IOleObject(iface);
+    IOleObject *inner;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    /* the viewer is the HTML document in front of the transformed XML: the call is for that document */
+    hres = IUnknown_QueryInterface(This->html_doc, &IID_IOleObject, (void**)&inner);
+    if(FAILED(hres))
+        return hres;
+    hres = IOleObject_Unadvise(inner, dwConnection);
+    IOleObject_Release(inner);
+    return hres;
 }
 
 static HRESULT WINAPI XMLView_OleObject_EnumAdvise(
         IOleObject *iface, IEnumSTATDATA **ppenumAdvise)
 {
     XMLView *This = impl_from_IOleObject(iface);
-    FIXME("(%p)->(%p)\n", This, ppenumAdvise);
-    return E_NOTIMPL;
+    IOleObject *inner;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    /* the viewer is the HTML document in front of the transformed XML: the call is for that document */
+    hres = IUnknown_QueryInterface(This->html_doc, &IID_IOleObject, (void**)&inner);
+    if(FAILED(hres))
+        return hres;
+    hres = IOleObject_EnumAdvise(inner, ppenumAdvise);
+    IOleObject_Release(inner);
+    return hres;
 }
 
 static HRESULT WINAPI XMLView_OleObject_GetMiscStatus(
         IOleObject *iface, DWORD dwAspect, DWORD *pdwStatus)
 {
-    FIXME("%p, %ld, %p.\n", iface, dwAspect, pdwStatus);
-    return E_NOTIMPL;
+    XMLView *This = impl_from_IOleObject(iface);
+    IOleObject *inner;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    /* the viewer is the HTML document in front of the transformed XML: the call is for that document */
+    hres = IUnknown_QueryInterface(This->html_doc, &IID_IOleObject, (void**)&inner);
+    if(FAILED(hres))
+        return hres;
+#ifdef SG_MUTANT_MSXML_XVOLE
+    hres = E_NOTIMPL;
+#else
+    hres = IOleObject_GetMiscStatus(inner, dwAspect, pdwStatus);
+#endif
+    IOleObject_Release(inner);
+    return hres;
 }
 
 static HRESULT WINAPI XMLView_OleObject_SetColorScheme(
         IOleObject *iface, LOGPALETTE *pLogpal)
 {
     XMLView *This = impl_from_IOleObject(iface);
-    FIXME("(%p)->(%p)\n", This, pLogpal);
-    return E_NOTIMPL;
+    IOleObject *inner;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    /* the viewer is the HTML document in front of the transformed XML: the call is for that document */
+    hres = IUnknown_QueryInterface(This->html_doc, &IID_IOleObject, (void**)&inner);
+    if(FAILED(hres))
+        return hres;
+    hres = IOleObject_SetColorScheme(inner, pLogpal);
+    IOleObject_Release(inner);
+    return hres;
 }
 
 static IOleObjectVtbl XMLView_OleObjectVtbl = {
